designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+languages that AWS Lambda supports. The cardinality of the problem is number of programming languages (N) times the
+number of package managers (M) per language. Hence is it is nearly impossible to natively support each combination.
+Instead, ``sam build`` will support an opinionated set of package managers for all programming languages. We will
+provide an option for customers to bring their own build commands or override the default for any programming language.
+SAM CLI will still take care of the grunt work of iterating through every function in the SAM template, figuring out
+the source code location, creating temporary folders to store built artifacts, running the build command and
+move artifacts to right location.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all programming languages supported by AWS Lambda
+
+   * Nodejs with NPM
+   * Java with Maven
+   * Python with PIP
+   * Golang with Go CLI
+   * Dotnetcore with DotNet CLI
+
+
+#. Each Lambda function in SAM template gets built
+
+#. Produce stable builds (best effort): If the source files did not change, built artifacts should not change.
+
+#. Built artifacts should "just work" with ``sam local`` and ``sam package`` suite of commands
+
+#. Ability to provide arbitrary build commands for each Lambda Function runtime. This will either override the built-in
+   default or add build support for languages/tools that SAM CLI does not support (ex: java+gradle).
+
+#. Opt-in to building native dependencies that can run on AWS Lambda using Docker.
+
+#. Support one dependency manifest (ex: package.json) entire app or one per each Lambda function.
+
+#. Support out-of-source builds: ie. source code of Lambda function is outside the directory containing SAM template.
+
+#. Integrate build action with ``sam local/package/deploy`` commands so the Lambda functions will be automatically
+   built as part of the command without explicitly running the build command.
+
+
+Out-of-Scope
+------------
+#. Supports adding data files ie. files that are not referenced by the package manager (ex: images, css etc)
+#. Support to exclude certain files from the built artifact (ex: using .gitignore or using regex)
+#. If the app contains a ``buildspec.yaml``, automatically run it using CodeBuild Local.
+#. Watch for file changes and build automatically (ex: ``sam build --watch``)
+#. Support other build systems by default Webpack, Yarn or Gradle.
+#. Support in AWS CLI, ``aws cloudformation``, suite of commands
+#. Support for fine-grained hooks (ex: hooks that run pre-build, post-build, etc)
+
+
+User Experience Walkthrough
+---------------------------
+Let's assume customers has the following SAM template:
+
+    **NOTE**: Currently we advice customers to set *CodeUri* to a folder containing built artifacts that can be readily
+    packaged by ``sam package`` command. But to use with *build* command, customers need to set *CodeUri* to the folder
```

Comment:
Customers do need to modify their SAM template to make use of build command (CodeUri points to source folder instead of built folder). Unfortunately there is no way around it. It is not a breaking change in the sense that your existing workflows won't break. Customers need to explicitly make this change to opt-into the build feature.

Also, the `built-template.yaml` is not the best UX. But besides introducing a `.sam` folder and storing these files there, I don't know of a good solution. Another option is to combine build & package so customers go straight to the `packaged-template.yaml` without the additional step.

Thoughts?